chalk-solve/src/lib.rs

```diff
@@ -36,6 +36,9 @@ pub trait RustIrDatabase<I: Interner>: Debug {
     /// Returns the `AssociatedTyValue` with the given id.
     fn associated_ty_value(&self, id: AssociatedTyValueId<I>) -> Arc<AssociatedTyValue<I>>;
 
+    /// Returns the `ImplTraitDatum` with the given id.
+    fn impl_trait_data(&self, id: ImplTraitId<I>) -> Arc<ImplTraitDatum<I>>;
```

Comment:
We don't have to do it in this PR, but I think eventually we will want to add a callback here like

```rust
fn hidden_type(&self, id: ImplTraitId<I>) -> Ty<I>;
```